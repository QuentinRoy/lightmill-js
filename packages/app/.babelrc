{
  "presets": [["env", {
    "targets": { "browsers": ">1%" },
    "modules": false,
    "loose": true
  }]],
  "plugins": ["external-helpers"]
}
