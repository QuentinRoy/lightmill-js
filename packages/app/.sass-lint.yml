rules:
  property-sort-order:
    - 1
    - order: 'concentric'
