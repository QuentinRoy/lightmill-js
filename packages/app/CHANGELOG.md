# Change Log

All notable changes to this project will be documented in this file.
See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="2.0.0-1"></a>
# 2.0.0-1 (2017-06-25)



<a name="2.0.0-0"></a>
# 2.0.0-0 (2017-06-25)


### Features

* **lightmill-app:** Statically export views. ([aa071b3](https://github.com/QuentinRoy/lightmill-js/tree/master/packages/lightmill-app/commit/aa071b3))




<a name="2.0.0-0"></a>
# 2.0.0-0 (2017-06-25)


### Features

* **lightmill-app:** Statically export views. ([aa071b3](https://github.com/QuentinRoy/lightmill-js/commit/aa071b3))
