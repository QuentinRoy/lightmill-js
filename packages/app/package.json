{
  "name": "@lightmill/app",
  "version": "2.0.0",
  "description": "Base application for lightmill-runner.",
  "main": "dist/lightmill-app.js",
  "files": [
    "dist"
  ],
  "devDependencies": {
    "babel-core": "6.26.3",
    "babel-plugin-external-helpers": "6.22.0",
    "babel-preset-env": "1.7.0",
    "cross-env": "5.1.6",
    "eslint": "4.19.1",
    "eslint-config-airbnb-base": "12.1.0",
    "eslint-plugin-import": "2.12.0",
    "jest-transform-pug": "0.1.0",
    "rollup": "0.59.3",
    "rollup-plugin-babel": "3.0.4",
    "rollup-plugin-commonjs": "9.1.3",
    "rollup-plugin-node-resolve": "3.3.0",
    "rollup-plugin-pug": "0.1.6",
    "rollup-plugin-sass": "0.7.0",
    "simulant": "0.2.2"
  },
  "dependencies": {
    "pug-runtime": "^2.0.3"
  },
  "scripts": {
    "test": "cross-env NODE_ENV=test jest",
    "coverage": "yarn test --coverage",
    "build": "cross-env NODE_ENV=production rollup --config rollup.config.js",
    "prepublish": "npm run build"
  },
  "keywords": [
    "experiment",
    "lightmill",
    "application"
  ],
  "author": "Quentin Roy <quentin@quentinroy.fr>",
  "license": "MIT",
  "ava": {
    "require": [
      "babel-register"
    ],
    "babel": "inherit"
  },
  "repository": "https://github.com/QuentinRoy/lightmill-js/tree/master/packages/lightmill-app",
  "jest": {
    "collectCoverageFrom": [
      "src/**/*.js"
    ],
    "transform": {
      "^.+\\.jsx?$": "babel-jest",
      "\\.(pug)$": "jest-transform-pug"
    },
    "moduleNameMapper": {
      "\\.(css|styl|less|sass|scss)$": "<rootDir>/../../test-tools/jest-stub-styles.js"
    }
  }
}
