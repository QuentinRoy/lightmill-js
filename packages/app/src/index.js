import end from './end';
import initBlock from './block-init';
import crash from './crash';
import wait from './wait';

/**
 * @typedef {Object} BlockInfo
 * @property {int} blockInfo.number The number of the block.
 * @property {int} blockInfo.measuredBlockNum The number of the block amongst
 * non practice blocks.
 * @property {boolean} blockInfo.practice If it is a practice block.
 * @property {Array<{factor:{name}, name}>} blockInfo.factorValues The list of
 * factor values.
 * @property {boolean} blockInfo.subjectiveAssessment If subjective assessment should be
 * asked for.
 */

/**
 * Standard Experimentation App Base class for lightmill-client providing wait, crash, end and
 * initBlock views. This is an abstract class that does not provide a `runTrial` handle
 * (required by xp-client). It is meant to be inherited (or monkey patched).
 * The app is highly reconfigurable and any method can be safely replaced at any time.
 * @constructor
 * @abstract
 */
class XpAppBase {
  /**
   * @param  {HTMLElement} node The node where to user for the app.
   */
  constructor(node) {
    /**
     * The node where lives the app (read-only).
     * @type {HTMLElement}
     */
    Object.defineProperty(this, 'node', { value: node });
  }

  /**
   * Must be overwritten by subclasses (or dynamically replaced).
   * @param {Object} config the trial configuration
   * @return {Promise<Object>} resolved with the trial results.
   * @abstract
   */
  runTrial() {
    throw new Error('runTrial not implemented');
  }

  /**
   * Show the wait view.
   * @returns {undefined}
   */
  wait() {
    wait(this.node);
  }

  /**
   * By default start waiting when the app starts.
   * @returns {undefined}
   */
  start() {
    this.wait();
  }

  /**
   * Show the crash view.
   * @param  {String} message The error message.
   * @param  {Error} [error]  The error that has been raised.
   * @param  {Object} [run]   The current run.
   * @returns {undefined}
   */
  crash(message, error, run) {
    crash(this.node, message, error, run);
  }

  /**
   * Show the block initialization view then show the wait view.
   * @param {BlockInfo} blockInfo Information about the block.
   * @return {Promise} Resolved when the user click/tap on the view.
   */
  initBlock(blockInfo) {
    return initBlock(this.node, blockInfo).then(res => {
      this.wait();
      return res;
    });
  }

  /**
   * Show the end view.
   * @returns {undefined}
   */
  end() {
    end(this.node);
  }
}

// Set up static the view handlers as static property of XpAppBase for
// cjs.
Object.assign(XpAppBase, { end, initBlock, crash, wait });

export { end, initBlock, crash, wait };
