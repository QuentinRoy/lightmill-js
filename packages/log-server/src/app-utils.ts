import type { paths } from '@lightmill/log-api';
import type { SessionData } from 'express-session';
import { groupBy, intersection, map, pipe, uniqueBy } from 'remeda';
import type { Simplify, WritableDeep } from 'type-fest';
import {
  type ApiPath,
  type HttpMethod,
  httpStatuses,
  type HttpStatusMap,
} from './api-utils.js';
import type { RunId, Store } from './store.js';
import type {
  Handler,
  HandlerParameters,
  HandlerResult,
  RequestContent,
  ServerDescription,
} from './typed-server.js';
import { arrayify } from './utils.js';

declare module 'express-session' {
  interface SessionData {
    data: { role: 'participant' | 'host'; runs: RunId[] };
  }
}

// This needs to be a type (not an interface) so we can use it with
// typed-server... I don't know why, but it's not worth investigating.
export type ServerApi = Simplify<paths>;

export type ServerHandler<
  Path extends ApiPath<ServerApi>,
  Method extends HttpMethod,
> = Handler<ServerApi, Path, Method>;

export type ServerHandlerResult<
  Path extends ApiPath<ServerApi>,
  Method extends HttpMethod,
> = HandlerResult<ServerApi, Path, Method>;

export type ServerHandlerBody<
  Path extends ApiPath<ServerApi>,
  Method extends HttpMethod,
> = ServerHandlerResult<Path, Method>['body'];

export type ServerHandlerIncluded<
  Path extends ApiPath<ServerApi>,
  Method extends HttpMethod,
> =
  ServerHandlerBody<Path, Method> extends infer B
    ? B extends { readonly included?: infer I }
      ? I
      : never
    : never;

export type ServerHandlerParameter<
  Path extends ApiPath<ServerApi>,
  Method extends HttpMethod,
> = HandlerParameters<ServerApi, Path, Method>;

export type ServerRequestContent<
  Path extends ApiPath<ServerApi>,
  Method extends HttpMethod,
> = RequestContent<ServerApi, Path, Method>;

export function getErrorResponse<
  const T extends {
    status: keyof HttpStatusMap;
    code: string;
    detail?: string;
    source?: object;
  },
>(option: T) {
  let error: WritableDeep<Omit<T, 'status'>> & {
    status: HttpStatusMap[T['status']];
  } = { ...structuredClone(option), status: httpStatuses[option.status] };
  return {
    status: option.status as T['status'],
    body: { errors: [error] },
    contentType: 'application/json' as const,
  };
}

export type SubServerDescription<K extends string> = Pick<
  ServerDescription<ServerApi>,
  Extract<keyof ServerDescription<ServerApi>, `${K}${string}`>
>;

type GetRunResourcesOptions =
  | { filter: Parameters<Store['getRuns']>[0] }
  | {
      runs: Omit<
        Awaited<ReturnType<Store['getRuns']>>[number],
        'runCreatedAt'
      >[];
    };
export async function getRunResources(
  store: Store,
  options: GetRunResourcesOptions,
) {
  const runs =
    'runs' in options ? options.runs : await store.getRuns(options.filter);
  const [experiments, lastLogs] = await Promise.all([
    store.getExperiments({
      experimentId: pipe(
        runs,
        uniqueBy((run) => run.experimentId),
        map((run) => run.experimentId),
      ),
    }),
    store.getLastLogs({ runId: runs.map((run) => run.runId) }),
  ]);
  const groupedLastLogs = groupBy(lastLogs, (log) => log.runId);

  return {
    runs: runs.map((run) => {
      const runLastLogs = groupedLastLogs[run.runId] ?? [];
      return {
        id: run.runId,
        type: 'runs' as const,
        attributes: {
          status: run.runStatus,
          name: run.runName,
          lastLogNumber: Math.max(0, ...runLastLogs.map((l) => l.number)),
        },
        relationships: {
          lastLogs: {
            data: runLastLogs.map((log) => ({
              id: log.logId,
              type: 'logs' as const,
            })),
          },
          experiment: {
            data: { id: run.experimentId, type: 'experiments' as const },
          },
        },
      };
    }),
    experiments: experiments.map((experiment) => ({
      id: experiment.experimentId,
      type: 'experiments' as const,
      attributes: { name: experiment.experimentName },
    })),
    lastLogs: lastLogs.map((log) => ({
      id: log.logId,
      type: 'logs' as const,
      attributes: { number: log.number, logType: log.type, values: log.values },
      relationships: {
        run: { data: { id: log.runId, type: 'runs' as const } },
      },
    })),
  };
}

export function getAllowedAndFilteredRunIds(
  sessionData: SessionData['data'] | undefined,
  queryFilter: undefined | string | string[],
) {
  if (sessionData == null) {
    return [];
  }
  if (sessionData.role === 'host') {
    return queryFilter;
  }
  if (queryFilter == null) {
    return sessionData.runs;
  }
  return intersection(sessionData.runs, arrayify(queryFilter, true));
}
