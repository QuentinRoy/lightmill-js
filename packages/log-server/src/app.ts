/* eslint-disable @typescript-eslint/no-namespace */
import express from 'express';
import { zodiosContext } from '@zodios/express';
import { LogFilter, Store } from './store.js';
import session from 'cookie-session';
import dotenv from 'dotenv';
import { SqliteError } from 'better-sqlite3';
import { NextFunction, Request, RequestHandler, Response } from 'express';
import { api } from './api.js';
import { csvExportStream, jsonExportStream } from './export.js';
import { pipeline } from 'stream/promises';
import log from 'loglevel';
import z from 'zod';
import cuid from 'cuid';

dotenv.config();

const ctx = zodiosContext(
  z.object({
    session: z.union([
      z.null(),
      z
        .object({
          role: z.union([z.literal('admin'), z.literal('participant')]),
          runs: z.array(
            z.object({
              runId: z.string(),
              experimentId: z.string(),
            })
          ),
        })
        .strict(),
    ]),
  })
);

type CreateAppParams = {
  store: Store;
  secret: string;
  adminPassword?: string;
};
export function createLogServer({
  store,
  secret,
}: CreateAppParams): RequestHandler {
  let app = express();
  const router = ctx.app(api, { express: app });

  router.use(session({ secret }));

  router.post('/sessions', (req, res) => {
    if (req.session?.role != null) {
      res.status(400).json({
        status: 'error',
        message: 'Client already has a session',
      });
      return;
    }
    const { role, password } = req.body;
    if (role === 'admin' && password !== process.env.ADMIN_PASSWORD) {
      res.status(403).json({
        status: 'error',
        message: `Forbidden role: ${role}`,
      });
    }
    req.session = { role, runs: [] };
    res.status(200).json({ status: 'ok', role, runs: [] });
  });

  router.get('/sessions/current', (req, res) => {
    if (req.session?.role == null) {
      res.status(404).json({
        status: 'error',
        message: 'No session found',
      });
      return;
    }
    res.status(200).json({
      status: 'ok',
      role: req.session.role,
      runs: req.session.runs.map((r) => ({
        id: r.runId,
        experiment: r.experimentId,
      })),
    });
  });

  router.delete('/sessions/current', (req, res) => {
    if (req.session?.role == null) {
      res.status(404).json({
        status: 'error',
        message: 'No session found',
      });
      return;
    }
    req.session = null;
    res.status(200).json({ status: 'ok' });
  });

  router.post('/experiments/runs', async (req, res, next) => {
    try {
      if (req.session?.role == null) {
        req.session = { role: 'participant', runs: [] };
      }
      if (req.session.runId != null) {
        res.status(403).json({
          status: 'error',
          message: 'Client already has a started run, end it first',
        });
        return;
      }
      let runId = req.body?.id ?? cuid();
      let experimentId = req.body?.experiment ?? 'default';
      const run = {
        ...req.body,
        experimentId,
        runId,
        createdAt: new Date(),
      };
      await store.addRun(run);
      req.session.runs.push({ runId, experimentId });
      res.status(200).json({
        status: 'ok',
        run: runId,
        experiment: experimentId,
        links: {
          logs: `/experiments/${experimentId}/runs/${runId}/logs`,
          run: `/experiments/${experimentId}/runs/${runId}`,
        },
      });
    } catch (e) {
      if (e instanceof SqliteError && e.code === 'SQLITE_CONSTRAINT') {
        res.status(400).json({
          status: 'error',
          message:
            'Could not add run, probably because a run with that ID already exists for this experiment',
        });
        return;
      }
      next(e);
    }
  });

  router.put('/experiments/:experiment/runs/:run', async (req, res, next) => {
    let { experiment: experimentId, run: runId } = req.params;
    experimentId = String(experimentId);
    runId = String(runId);
    try {
      if (
        req.session?.runs?.find(
          (r) => r.runId === runId && r.experimentId === experimentId
        ) == null
      ) {
        res.status(403).json({
          status: 'error',
          message: `Client does not have permission to update run "${runId}" of experiment "${experimentId}"`,
        });
        return;
      }
      let run = await store.getRun(experimentId, runId);
      if (run == null) {
        throw new Error(`Session run not found: ${runId}`);
      }

      // At the moment, the only supported put operation is
      // { "status": "completed" | "canceled" }, so there is nothing more to
      // check here, zodios does it for us already.
      if (run.status != 'running') {
        res.status(400).json({
          status: 'error',
          message: 'Run already ended',
        });
        return;
      }
      await store.setRunStatus(experimentId, runId, req.body.status);
      req.session.runs = req.session.runs.filter(
        (r) => r.runId !== runId && r.experimentId !== experimentId
      );
      res.status(200).json({ status: 'ok' });
    } catch (e) {
      next(e);
    }
  });

  router.post(
    '/experiments/:experiment/runs/:run/logs',
    async (req, res, next) => {
      let { experiment: experimentId, run: runId } = req.params;
      experimentId = String(experimentId);
      runId = String(runId);
      try {
        if (
          req.session?.runs?.find(
            (r) => r.runId === runId && r.experimentId === experimentId
          ) == null
        ) {
          res.status(403).json({
            status: 'error',
            message: `Client does not have permission to post logs for run "${runId}" of experiment "${experimentId}"`,
          });
          return;
        }
        let sessionRun = await store.getRun(experimentId, runId);
        if (sessionRun == null) {
          // This should not happen in normal use, except if the database is
          // corrupted, or removed.
          throw new Error(`Session run not found: ${runId}`);
        }
        if (sessionRun.status != 'running') {
          res.status(403).json({
            status: 'error',
            message: 'Cannot add logs to an ended run',
          });
          return;
        }
        let logs = 'logs' in req.body ? req.body.logs : [req.body.log];
        await store.addRunLogs(experimentId, runId, logs);
        res.status(200).json({ status: 'ok' });
      } catch (e) {
        next(e);
      }
    }
  );

  router.get('/experiments/:experiment/runs/logs', async (req, res, next) => {
    try {
      if (req.session?.role !== 'admin') {
        res.status(403).json({
          status: 'error',
          message: 'Access restricted.',
        });
        return;
      }
      let { format, type } = req.query;
      let { experiment } = req.params ?? {};
      let filter: LogFilter = {
        experiment: experiment == null ? undefined : String(experiment),
        type: type,
      };
      if (format === 'csv') {
        res.setHeader('Content-Type', 'text/csv');
        await pipeline(csvExportStream(store, filter), res);
      } else {
        res.setHeader('Content-Type', 'application/json');
        await pipeline(jsonExportStream(store, filter), res);
      }
    } catch (e) {
      next(e);
    }
  });

  router.use('*', (req, res) => {
    res.status(404).json({
      status: 'error',
      message: 'Not found',
    });
  });

  app.use((error: unknown, req: Request, res: Response, next: NextFunction) => {
    if (res.headersSent) {
      next(error);
    } else if (error instanceof Error) {
      res.status(500).json({
        status: 'error',
        message: error.message,
      });
    } else {
      log.error('error', error);
      next(error);
    }
  });

  return app;
}
