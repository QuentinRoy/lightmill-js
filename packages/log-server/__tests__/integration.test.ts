/* eslint-disable no-empty-pattern */
import express from 'express';
import request from 'supertest';
import {
  afterEach,
  describe,
  expect,
  vi,
  test as vitestTest,
  type TestAPI,
} from 'vitest';
import { apiMediaType } from '../src/app-utils.js';
import { LogServer } from '../src/app.js';
import { SQLiteStore } from '../src/store.js';
import { apiContentTypeRegExp } from './test-utils.js';

// Reset after each test
afterEach(() => {
  vi.resetAllMocks();
});

// Define fixture context types
type FixtureContext = { api: request.Agent };
type App = NonNullable<Parameters<typeof request.agent>[0]>;
type Fixture = {
  app: App;
  hostSession: FixtureContext & { role: 'host' };
  participantSession: FixtureContext & { role: 'participant' };
};

// Create fixtures using vitest's extend
const baseTest: TestAPI<Fixture> = vitestTest.extend<Fixture>({
  app: async ({}, use) => {
    // Create real in-memory database
    const store = new SQLiteStore(':memory:');
    await store.migrateDatabase();
    // Server setup with real stores
    const server = LogServer({
      store,
      sessionKeys: ['integration-test-secret'],
      hostUser: 'admin',
      hostPassword: 'password',
      secureCookies: false,
    });
    // Express app setup
    const app = express().use(server.middleware);
    await use(app);
    // Clean up after test
    await store.close();
  },

  // Host session fixture
  hostSession: async ({ app }, use) => {
    // Create authenticated host agent
    const api = request.agent(app);
    await api
      .post('/sessions')
      .auth('admin', 'password')
      .set('Content-Type', apiMediaType)
      .send({ data: { type: 'sessions', attributes: { role: 'host' } } })
      .expect(201);

    await use({ api, role: 'host' });
  },

  // Participant session fixture
  participantSession: async ({ app }, use) => {
    // Create authenticated participant agent
    const api = request.agent(app);
    await api
      .post('/sessions')
      .set('Content-Type', apiMediaType)
      .send({ data: { type: 'sessions', attributes: { role: 'participant' } } })
      .expect(201);

    await use({ api, role: 'participant' });
  },
});

describe('Integration - Authentication', () => {
  const it = baseTest.extend<{ agent: request.Agent }>({
    agent: async ({ app }, use) => {
      const agent = request.agent(app);
      await use(agent);
    },
  });

  it('creates a valid participant session', async ({ agent }) => {
    const response = await agent
      .post('/sessions')
      .set('Content-Type', apiMediaType)
      .send({ data: { type: 'sessions', attributes: { role: 'participant' } } })
      .expect(201);

    expect(response.body.data.attributes.role).toBe('participant');
  });

  it('creates a host session with valid credentials', async ({ agent }) => {
    const response = await agent
      .post('/sessions')
      .auth('admin', 'password')
      .set('Content-Type', apiMediaType)
      .send({ data: { type: 'sessions', attributes: { role: 'host' } } })
      .expect(201);

    expect(response.body.data.attributes.role).toBe('host');
  });

  it('rejects host creation with invalid credentials', async ({ agent }) => {
    await agent
      .post('/sessions')
      .auth('admin', 'wrong-password')
      .set('Content-Type', apiMediaType)
      .send({ data: { type: 'sessions', attributes: { role: 'host' } } })
      .expect(403);
  });
});

describe('Integration - Experiment Management', () => {
  const it = baseTest;
  it('creates and retrieves experiments as host', async ({
    hostSession: { api },
  }) => {
    // Create experiment
    const experimentName = `test-experiment`;
    const createResponse = await api
      .post('/experiments')
      .set('Content-Type', apiMediaType)
      .send({
        data: { type: 'experiments', attributes: { name: experimentName } },
      })
      .expect(201)
      .expect('Content-Type', apiContentTypeRegExp);

    const experimentId = createResponse.body.data.id;
    expect(experimentId).toBeDefined();

    // Get experiment
    const getResponse = await api
      .get(`/experiments/${experimentId}`)
      .expect(200)
      .expect('Content-Type', apiContentTypeRegExp);

    expect(getResponse.body.data.attributes.name).toBe(experimentName);

    // List experiments
    const listResponse = await api
      .get('/experiments')
      .expect(200)
      .expect('Content-Type', apiContentTypeRegExp);

    expect(listResponse.body.data).toBeInstanceOf(Array);
    expect(listResponse.body.data).toSatisfy((data: Array<{ id: unknown }>) =>
      data.some((exp: { id: unknown }) => exp.id === experimentId),
    );
  });

  it('prevents participants from creating experiments', async ({
    participantSession: { api },
  }) => {
    await api
      .post('/experiments')
      .set('Content-Type', apiMediaType)
      .send({
        data: {
          type: 'experiments',
          attributes: { name: 'participant-experiment' },
        },
      })
      .expect(403);
  });

  it('prevents duplicate experiment names', async ({
    hostSession: { api },
  }) => {
    const experimentName = `unique-experiment`;

    // Create first experiment
    await api
      .post('/experiments')
      .set('Content-Type', apiMediaType)
      .send({
        data: { type: 'experiments', attributes: { name: experimentName } },
      })
      .expect(201);

    // Try to create experiment with same name
    await api
      .post('/experiments')
      .set('Content-Type', apiMediaType)
      .send({
        data: { type: 'experiments', attributes: { name: experimentName } },
      })
      .expect(409);
  });
});

describe('Integration - Run Management', () => {
  // Create a fixture with experiment pre-created
  const it = baseTest.extend<{
    hostSession: FixtureContext & { experimentId: string };
    participantSession: FixtureContext & { experimentId: string };
  }>({
    hostSession: async ({ hostSession }, use) => {
      const experimentName = `run-test-experiment`;
      const createResponse = await hostSession.api
        .post('/experiments')
        .set('Content-Type', apiMediaType)
        .send({
          data: { type: 'experiments', attributes: { name: experimentName } },
        })
        .expect(201);

      const experimentId = createResponse.body.data.id;
      await use({ ...hostSession, experimentId });
    },
    participantSession: async ({ participantSession, hostSession }, use) => {
      const experimentId = hostSession.experimentId;
      await use({ ...participantSession, experimentId });
    },
  });

  it('creates, updates and retrieves runs', async ({
    hostSession: { api, experimentId },
  }) => {
    // Create run
    const runName = `test-run`;
    const createResponse = await api
      .post('/runs')
      .set('Content-Type', apiMediaType)
      .send({
        data: {
          type: 'runs',
          attributes: { name: runName, status: 'idle' },
          relationships: {
            experiment: { data: { type: 'experiments', id: experimentId } },
          },
        },
      })
      .expect(201)
      .expect('Content-Type', apiContentTypeRegExp);

    const runId = createResponse.body.data.id;
    expect(runId).toBeDefined();

    // Get run
    const getResponse = await api
      .get(`/runs/${runId}`)
      .expect(200)
      .expect('Content-Type', apiContentTypeRegExp);

    expect(getResponse.body.data.attributes.name).toBe(runName);
    expect(getResponse.body.data.attributes.status).toBe('idle');

    // Update run status
    await api
      .patch(`/runs/${runId}`)
      .set('Content-Type', apiMediaType)
      .send({
        data: { id: runId, type: 'runs', attributes: { status: 'running' } },
      })
      .expect(200);

    // Verify status update
    const updatedResponse = await api.get(`/runs/${runId}`).expect(200);

    expect(updatedResponse.body.data.attributes.status).toBe('running');
    expect(updatedResponse.body).toMatchSnapshot();
  });

  it('adds run to participant session', async ({
    participantSession: { api, experimentId },
  }) => {
    // Create run
    const createResponse = await api
      .post('/runs')
      .set('Content-Type', apiMediaType)
      .send({
        data: {
          type: 'runs',
          attributes: { name: `session-test-run`, status: 'idle' },
          relationships: {
            experiment: { data: { type: 'experiments', id: experimentId } },
          },
        },
      })
      .expect(201);

    const runId = createResponse.body.data.id;
    expect(runId).toBeDefined();

    // Now participant should have access
    const runResponse = await api.get(`/runs/${runId}`).expect(200);

    expect(runResponse.body.data.id).toBe(runId);
    expect(runResponse.body).toMatchSnapshot();
  });
});

describe('Integration - Logging Flow', () => {
  // Create a fixture with experiment and run pre-created
  const it = baseTest.extend<{
    hostSession: FixtureContext & { experimentId: string; runId: string };
    participantSession: FixtureContext & {
      experimentId: string;
      runId: string;
    };
  }>({
    hostSession: async ({ hostSession }, use) => {
      // Create experiment
      const experimentName = `log-test-experiment`;
      const experimentResponse = await hostSession.api
        .post('/experiments')
        .set('Content-Type', apiMediaType)
        .send({
          data: { type: 'experiments', attributes: { name: experimentName } },
        })
        .expect(201);

      const experimentId = experimentResponse.body.data.id;

      // Create run
      const runName = `log-test-host-run`;
      const runResponse = await hostSession.api
        .post('/runs')
        .set('Content-Type', apiMediaType)
        .send({
          data: {
            type: 'runs',
            attributes: { name: runName, status: 'running' },
            relationships: {
              experiment: { data: { type: 'experiments', id: experimentId } },
            },
          },
        })
        .expect(201);

      const runId = runResponse.body.data.id;
      expect(runId).toBeTypeOf('string');

      await use({ ...hostSession, experimentId, runId });
    },
    participantSession: async ({ participantSession, hostSession }, use) => {
      const experimentId = hostSession.experimentId;
      const runName = `log-test-participant-run`;

      // Add run to participant's session
      const resp = await participantSession.api
        .post('/runs')
        .set('Content-Type', apiMediaType)
        .send({
          data: {
            type: 'runs',
            attributes: { name: runName, status: 'running' },
            relationships: {
              experiment: { data: { type: 'experiments', id: experimentId } },
            },
          },
        })
        .expect(201);

      const runId = resp.body.data.id;
      expect(runId).toBeTypeOf('string');

      await use({
        ...participantSession,
        experimentId: hostSession.experimentId,
        runId,
      });
    },
  });

  it('creates and retrieves logs', async ({
    participantSession: { api, runId },
  }) => {
    // Add logs
    await api
      .post(`/logs`)
      .set('Content-Type', apiMediaType)
      .send({
        data: {
          type: 'logs',
          attributes: { number: 1, logType: 'test', values: { x: 'x' } },
          relationships: { run: { data: { type: 'runs', id: runId } } },
        },
      })
      .expect(201)
      .expect('Content-Type', apiContentTypeRegExp);

    await api
      .post(`/logs`)
      .set('Content-Type', apiMediaType)
      .send({
        data: {
          type: 'logs',
          attributes: { number: 2, logType: 'test', values: { x: 'y' } },
          relationships: { run: { data: { type: 'runs', id: runId } } },
        },
      })
      .expect(201)
      .expect('Content-Type', apiContentTypeRegExp);

    // Get logs
    const logsResponse = await api.get(`/logs`).expect(200);

    expect(logsResponse.text).toMatchInlineSnapshot(`
      "type,experiment_name,run_name,run_status,x
      test,log-test-experiment,log-test-participant-run,running,x
      test,log-test-experiment,log-test-participant-run,running,y
      "
    `);
  });

  it('filters logs by type', async ({ participantSession: { api, runId } }) => {
    // Add logs of different types

    async function postLog(logType: string, number: number, values: object) {
      return api
        .post(`/logs`)
        .set('Content-Type', apiMediaType)
        .send({
          data: {
            type: 'logs',
            attributes: { number, logType, values },
            relationships: { run: { data: { type: 'runs', id: runId } } },
          },
        })
        .expect(201);
    }

    await postLog('click', 1, { clickValue: 'x' });
    await postLog('move', 2, { moveValue: 'y' });
    await postLog('click', 3, { clickValue: 'z' });

    // Get filtered logs
    const logsResponse = await api
      .get(`/logs?filter[logType]=click`)
      .expect(200);

    expect(logsResponse.text).toMatchInlineSnapshot(`
      "experiment_name,run_name,run_status,click_value
      log-test-experiment,log-test-participant-run,running,x
      log-test-experiment,log-test-participant-run,running,z
      "
    `);
  });

  it('handles complex log value types', async ({
    participantSession: { api, runId },
  }) => {
    // Add log with complex values
    await api
      .post(`/logs`)
      .set('Content-Type', apiMediaType)
      .send({
        data: {
          type: 'logs',
          relationships: { run: { data: { type: 'runs', id: runId } } },
          attributes: {
            number: 1,
            logType: 'complex',
            values: {
              nested: { a: 1, b: 'string' },
              array: [1, 2, 3],
              object_array: [{ id: 1 }, { id: 2 }],
              null_value: null,
              boolean: true,
            },
          },
        },
      })
      .expect(201);

    // Get log and verify complex structure
    const logsResponse = await api
      .get(`/logs`)
      .set('Accept', apiMediaType)
      .expect(200);
    expect(logsResponse.body).toMatchSnapshot();
  });

  it('tests run resumption with logs', async ({
    hostSession: { api, runId },
  }) => {
    // Add initial logs
    // Log 1
    await api
      .post(`/logs`)
      .set('Content-Type', apiMediaType)
      .send({
        data: {
          type: 'logs',
          attributes: { number: 1, logType: 'test', values: { value: 'log1' } },
          relationships: { run: { data: { type: 'runs', id: runId } } },
        },
      })
      .expect(201);

    // Log 2
    await api
      .post(`/logs`)
      .set('Content-Type', apiMediaType)
      .send({
        data: {
          type: 'logs',
          attributes: { number: 2, logType: 'test', values: { value: 'log2' } },
          relationships: { run: { data: { type: 'runs', id: runId } } },
        },
      })
      .expect(201);

    // Log 3
    await api
      .post(`/logs`)
      .set('Content-Type', apiMediaType)
      .send({
        data: {
          type: 'logs',
          attributes: { number: 3, logType: 'test', values: { value: 'log3' } },
          relationships: { run: { data: { type: 'runs', id: runId } } },
        },
      })
      .expect(201);

    let response = await api.get(`/logs`).expect(200);
    expect(response.text).toMatchInlineSnapshot(`
      "type,experiment_name,run_name,run_status,value
      test,log-test-experiment,log-test-host-run,running,log1
      test,log-test-experiment,log-test-host-run,running,log2
      test,log-test-experiment,log-test-host-run,running,log3
      "
    `);

    // Set run to interrupted
    await api
      .patch(`/runs/${runId}`)
      .set('Content-Type', apiMediaType)
      .send({
        data: {
          id: runId,
          type: 'runs',
          attributes: { status: 'interrupted' },
        },
      })
      .expect(200);

    // Resume from log 2
    await api
      .patch(`/runs/${runId}`)
      .set('Content-Type', apiMediaType)
      .send({
        data: {
          id: runId,
          type: 'runs',
          attributes: { status: 'running', lastLogNumber: 2 },
        },
      })
      .expect(200);

    // Add new log that should replace log 3
    await api
      .post(`/logs`)
      .set('Content-Type', apiMediaType)
      .send({
        data: {
          type: 'logs',
          attributes: {
            number: 3,
            logType: 'test',
            values: { value: 'log3-replaced' },
          },
          relationships: { run: { data: { type: 'runs', id: runId } } },
        },
      })
      .expect(201);

    // Verify logs
    response = await api.get(`/logs`).expect(200);
    expect(response.text).toMatchInlineSnapshot(`
      "type,experiment_name,run_name,run_status,value
      test,log-test-experiment,log-test-host-run,running,log1
      test,log-test-experiment,log-test-host-run,running,log2
      test,log-test-experiment,log-test-host-run,running,log3-replaced
      "
    `);
  });
});
