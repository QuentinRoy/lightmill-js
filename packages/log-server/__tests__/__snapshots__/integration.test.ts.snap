// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Integration - Logging Flow > handles complex log value types 1`] = `
{
  "data": [
    {
      "attributes": {
        "logType": "complex",
        "number": 1,
        "values": {
          "array": [
            1,
            2,
            3,
          ],
          "boolean": true,
          "nested": {
            "a": 1,
            "b": "string",
          },
          "null_value": null,
          "object_array": [
            {
              "id": 1,
            },
            {
              "id": 2,
            },
          ],
        },
      },
      "id": "1",
      "relationships": {
        "run": {
          "data": {
            "id": "2",
            "type": "runs",
          },
        },
      },
      "type": "logs",
    },
  ],
}
`;

exports[`Integration - Run Management > adds run to participant session 1`] = `
{
  "data": {
    "attributes": {
      "lastLogNumber": 0,
      "name": "session-test-run",
      "status": "idle",
    },
    "id": "1",
    "relationships": {
      "experiment": {
        "data": {
          "id": "1",
          "type": "experiments",
        },
      },
      "lastLogs": {
        "data": [],
      },
    },
    "type": "runs",
  },
}
`;

exports[`Integration - Run Management > creates, updates and retrieves runs 1`] = `
{
  "data": {
    "attributes": {
      "lastLogNumber": 0,
      "name": "test-run",
      "status": "running",
    },
    "id": "1",
    "relationships": {
      "experiment": {
        "data": {
          "id": "1",
          "type": "experiments",
        },
      },
      "lastLogs": {
        "data": [],
      },
    },
    "type": "runs",
  },
}
`;
